File: code/ex143.java
import java.util.*;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

class Solution {
    public void reorderList(ListNode head) {
        if (head == null || head.next == null)
            return;
        
        // Save nodes to Deque
        Deque<ListNode> deque = new LinkedList<>();
        for (ListNode p = head.next; p != null; p = p.next)
            deque.add(p);
        
        // Insert
        head.next = null;

        ListNode p = head;       
        while (true) {
            p.next = deque.pollLast();  
            p = p.next;
            if (deque.isEmpty()) break;
            
            p.next = deque.pollFirst();
            p = p.next;
            if (deque.isEmpty()) break;
        }
        
        p.next = null;
    }
}

File: code/ex143.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        vector<ListNode*> nodes;
        for (ListNode* p = head; p != NULL; p = p->next)
            nodes.push_back(p);
        
        int n = nodes.size();
        for (int i = 0, j = n - 1; i < (n - 1) / 2; i++, j--) {
            nodes[j]->next = nodes[i]->next;
            nodes[i]->next = nodes[j];
            nodes[j - 1]->next = NULL;
        }
    }
};